src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverterSettings.cs

```diff
@@ -0,0 +1,235 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+
+namespace System.Text.Json.Serialization
+{
+    public class JsonConverterSettings
+    {
+        private const int DefaultBufferSizeInternal = 16 * 1024;
+
+        private volatile JsonMemberBasedClassMaterializer _classMaterializerStrategy;
+        private JsonClassMaterializer _classMaterializer;
+        private int _defaultBufferSize = DefaultBufferSizeInternal;
+        private int _maxDepth = 64;
+        private bool _hasRuntimeCustomAttributes;
+
+        private static readonly Dictionary<ICustomAttributeProvider, object[]> s_reflectionAttributes = new Dictionary<ICustomAttributeProvider, object[]>();
+        private readonly Lazy<Dictionary<ICustomAttributeProvider, List<Attribute>>> _runtimeAttributes = new Lazy<Dictionary<ICustomAttributeProvider, List<Attribute>>>();
+
+        private static readonly ConcurrentDictionary<Type, JsonClassInfo> s_classes = new ConcurrentDictionary<Type, JsonClassInfo>();
+        private readonly ConcurrentDictionary<Type, JsonClassInfo> _local_classes = new ConcurrentDictionary<Type, JsonClassInfo>();
+
+        //todo: exception once used on all globals
+
+        internal JsonClassInfo GetOrAddClass(Type classType)
+        {
+            JsonClassInfo result;
+
+            // Once custom attributes have been used, cache classes locally.
+            if (_hasRuntimeCustomAttributes)
+            {
+                if (!_local_classes.TryGetValue(classType, out result))
+                {
+                    result = _local_classes.GetOrAdd(classType, new JsonClassInfo(classType, this));
+                }
+            }
+            else
+            {
+                if (!s_classes.TryGetValue(classType, out result))
+                {
+                    result = s_classes.GetOrAdd(classType, new JsonClassInfo(classType, this));
+                }
+            }
+
+            return result;
+        }
+
+        public JsonReaderOptions ReaderOptions { get; set; }
+
+        public JsonClassMaterializer ClassMaterializer
+        {
+            get
+            {
+                return _classMaterializer;
+            }
+            set
+            {
+                if (_classMaterializer != JsonClassMaterializer.Default && value != _classMaterializer)
+                {
+                    throw new InvalidOperationException("Can't change materializer");
+                }
+
+                _classMaterializer = value;
+            }
+        }
+
+        public int DefaultBufferSize
+        {
+            get
+            {
+                return _defaultBufferSize;
+            }
+            set
+            {
+                if (value < 1)
+                {
+                    throw new ArgumentException("todo");
+                }
```

Comment:
Which style do you use here and in some other places? I see that somewhere you have brackets around one line statements, but not everywhere. Maybe remove them where needed and keep code cleaner?